File: frontend/src/lib/vehicles.ts
import { api } from "./api";

export interface VehicleOut {
  id: string;
  tenant_id: string;
  fonte: string;
  tipo: string;
  marca: string;
  modelo: string;
  ano_modelo: number;
  combustivel: string | null;
  codigo_fipe: string | null;
  valor_fipe: string | null;
  valor_referencia: string | null;
  mes_referencia_fipe: string | null;
  cor: string | null;
  placa: string | null;
  odometro_km: number | null;
  status: "ativo" | "reservado" | "vendido" | "inativo";
  snapshot_json: Record<string, unknown> | null;
  criado_por: string;
  criado_em: string;
  atualizado_em: string;
}

export interface VehicleListPage {
  items: VehicleOut[];
  next_cursor: string | null;
}

export interface VehicleIn {
  fonte: string;
  tipo: string;
  marca: string;
  modelo: string;
  ano_modelo: number;
  combustivel?: string | null;
  codigo_fipe?: string | null;
  valor_fipe?: string | null;
  valor_referencia?: string | null;
  mes_referencia_fipe?: string | null;
  cor?: string | null;
  placa?: string | null;
  odometro_km?: number | null;
  snapshot_json?: Record<string, unknown> | null;
}

export async function listVehicles(params?: { status?: string; placa?: string; cursor?: string }): Promise<VehicleListPage> {
  const { data } = await api.get<VehicleListPage>("/api/v1/vehicles", { params });
  return data;
}

export async function createVehicle(body: VehicleIn): Promise<VehicleOut> {
  const { data } = await api.post<VehicleOut>("/api/v1/vehicles", body);
  return data;
}

export async function getVehicle(id: string): Promise<VehicleOut> {
  const { data } = await api.get<VehicleOut>(`/api/v1/vehicles/${id}`);
  return data;
}

export async function setVehicleStatus(id: string, status: string): Promise<VehicleOut> {
  const { data } = await api.post<VehicleOut>(`/api/v1/vehicles/${id}/status`, { status });
  return data;
}

export async function refreshVehicleFipe(id: string): Promise<VehicleOut> {
  const { data } = await api.post<VehicleOut>(`/api/v1/vehicles/${id}/refresh-fipe`);
  return data;
}

File: backend/finacialsim_saas/schemas/vehicles.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel

from finacialsim_saas.schemas.types import DecimalStr


class VehicleIn(BaseModel):
    fonte: str         # "fipe_parallelum" | "fipe_brasilapi" | "manual"
    tipo: str          # "carro" | "moto" | "caminhao"
    marca: str
    modelo: str
    ano_modelo: int
    combustivel: str | None = None
    codigo_fipe: str | None = None
    valor_fipe: DecimalStr | None = None
    valor_referencia: DecimalStr | None = None
    mes_referencia_fipe: str | None = None
    cor: str | None = None
    placa: str | None = None
    odometro_km: int | None = None
    snapshot_json: dict | None = None


class VehicleOut(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    fonte: str
    tipo: str
    marca: str
    modelo: str
    ano_modelo: int
    combustivel: str | None
    codigo_fipe: str | None
    valor_fipe: DecimalStr | None
    valor_referencia: DecimalStr | None
    mes_referencia_fipe: str | None
    cor: str | None
    placa: str | None
    odometro_km: int | None
    status: str
    snapshot_json: dict | None
    criado_por: uuid.UUID
    criado_em: datetime
    atualizado_em: datetime

    class Config:
        from_attributes = True


class VehicleListItem(BaseModel):
    id: uuid.UUID
    tipo: str
    marca: str
    modelo: str
    ano_modelo: int
    placa: str | None
    valor_fipe: DecimalStr | None
    status: str
    criado_em: datetime

    class Config:
        from_attributes = True


class VehicleListPage(BaseModel):
    items: list[VehicleListItem]
    next_cursor: str | None


class VehicleStatusUpdate(BaseModel):
    status: str
